src/qt/diagnosticsdialog.cpp

```diff
@@ -0,0 +1,371 @@
+#include <QtConcurrent>
+
+#include <fstream>
+#include "main.h"
+#include "util.h"
+#include "boinc.h"
+#include <boost/algorithm/string.hpp>
+#include <boost/filesystem.hpp>
+#include <boost/date_time/posix_time/posix_time.hpp>
+
+#include "diagnosticsdialog.h"
+#include "ui_diagnosticsdialog.h"
+
+std::string GetListOf(std::string datatype);
+double PreviousBlockAge();
+
+DiagnosticsDialog::DiagnosticsDialog(QWidget *parent) :
+    QDialog(parent),
+    ui(new Ui::DiagnosticsDialog)
+{
+    ui->setupUi(this);
+
+    networkManager = new QNetworkAccessManager(this);
+    connect(networkManager, SIGNAL(finished(QNetworkReply *)), this, SLOT(getGithubVersionFinished(QNetworkReply *)));
+}
+
+DiagnosticsDialog::~DiagnosticsDialog()
+{
+    delete ui;
+}
+
+void DiagnosticsDialog::GetData() {
+    std::string cpiddata = GetListOf("beacon");
```

Comment:
I think we should try to remove the GetData() method and instead move each part to where the check is. For example, move the `std::string cpiddata = GetListOf("beacon");` call down to `VerifyCPIDIsInNeuralNetwork()`. That way we can skip the ExtractXML where possible.

It's probably a remnant from when we had to assemble a data string to send to VB, right?

Reply:
Yeah that's right. Cool, yeah makes sense.